### ISSUE TITLE
[Bug] Livestreams don't start when they start on YouTube

### ISSUE THREAD

**Issue by hugoamvieira** [iv-org/invidious#2645]
**Describe the bug**
As I stand here, it took around 11mins for a livestream I was waiting on to start on my Invidious instance.

The livestream started at 4pm on YouTube, and it was only 'til 4:11pm that Invidious picked up on the feed that had been going since 4pm on YouTube's website. During these 11 minutes, I tried restarting Invidious, hard refreshed the page multiple times and the only thing I ever got was a page with the stream thumbnail and a message "This live event will begin in a few moments."

In terms of errors, I'm afraid I got nothing here. It's also happened before at least 3x that I can remember.

Additionally, sometimes these events don't even show up in the subscription feed (or the actual channel page) when you're subscribed to the channel.

**Steps to Reproduce**
1. Go on Invidious and open a livestream (maybe premiere too?) that will start in the future;
2. Open it on YouTube too
3. Observe that stream starts on YouTube
4. Stream doesn't start on Invidious
5. Refresh Invidious page - Stream still doesn't start.

**Logs**
N/A

**Screenshots**
N/A

**Additional context**
Invidious 2021.11.16-e0d26f4
Firefox 94.0.1

**Comment by unixfox** [comment]
If you don't have pubsub enabled then it's normal to not have the livestream in the feed directly, invidious refresh all the channels that you subscribed every 5 seconds but if it has a lot of channels to refresh then it may take some time.

And I don't fully understand but no livestreams are working for you?

**Comment by hugoamvieira** [comment]
> If you don't have pubsub enabled then it's normal to not have the livestream in the feed directly

Just to make sure I fully understand what you mean here - Do you mean if I don't use pubsub then it's normal for livestreams to not show up in my subscription feed?
I'm having a tough time understanding this since in my instance sometimes they _do_ show up, and sometimes they don't.

> And I don't fully understand but no livestreams are working for you?

Sorry if my explanation was unclear. They **are** working. I can watch them just fine.

The problem I see is with scheduled livestreams where Invidious is waiting for a stream to go live and the timing around that - in my experience, it's way behind YouTube - that is to say that when my instance detects that the stream is live and starts playing, it will have been live on YouTube for several minutes already - in my experience, in the order of tens of minutes later.

My Invidious instance has 1 user (me) with 101 subscriptions. I don't think I can tell how long it takes to refresh all of that - I see nothing in the debug logging - there are references to `RefreshChannelsJob` which seems to take care of refreshing channels themselves (?) and that seems to be taking about 3-4 seconds start to finish.

Pertinent config (I think)
```
channel_threads: 16
feed_threads: 16
```

**Comment by unixfox** [comment]
> Just to make sure I fully understand what you mean here - Do you mean if I don't use pubsub then it's normal for livestreams to not show up in my subscription feed?

Do you have pubsub enabled: https://github.com/iv-org/invidious/blob/master/config/config.example.yml#L408?

**Comment by hugoamvieira** [comment]
> Do you have pubsub enabled

Nope

**Comment by unixfox** [comment]
> > Do you have pubsub enabled
> 
> Nope

Then you should try if that would fix the issue where the livestream doesn't appear immediately in your feeds.

**Comment by unixfox** [comment]
Any update on the matter?

**Comment by hugoamvieira** [comment]
Hey @unixfox thanks for checking in - I haven't had time to work on this however I did want to pick your brains a little bit about best practices.

I don't actually need my instance to be exposed to the world in terms of me accessing it, and exposing it to the internet would increase risk, so I'd like to be able to lock it down such that only me (as someone in the LAN) and the pubsub service can talk to it.

I was thinking of perhaps using IPs as a way of restricting this. Block all IPs and then whitelist specific IPs/CIDRs.

What are your thoughts on this? Thanks!

**Comment by unixfox** [comment]
Blocking all the IPs then whitelist only the IP ranges from Google is a good idea indeed.

Or you can just block all the paths and just whitelist the path `/feed/webhook` which is the only path used by google pubsub.

I think the latter one is probably the easiest to set up and maintain.

**Comment by hugoamvieira** [comment]
@unixfox thanks a lot! I'm going to give this a shot and then report back

**Comment by hugoamvieira** [comment]
@unixfox seems to work well - 
The only problem I came across was the requirement to set the `domain` field - Because I had to set that to my external hostname, but that only exposed `/feed/webhook/v1/`, watching livestreams didn't work (due to CSP problems since there was a hostname mismatch) when I accessed it through my internal hostname.

**Comment by unixfox** [comment]
Why aren't you using the domain that you created? If you need to access it from your internal network just override the IP address in your hosts file.

**Comment by hugoamvieira** [comment]
I think I discovered something new that while is not related with livestreams, it is related with videos showing up in the subscription feed, and that might include livestreams (not sure).

This happened just now and I found it by comparing my feed from FreeTube to the Invidious instance. My FreeTube feed showed this video from `bigclivedotcom`, and my invidious didn't.

First thing I did was go directly into the channel, and the video was there:
![in-channel](https://user-images.githubusercontent.com/10965479/145696813-8ab649cf-be9e-42e0-8bd0-a29a12d6ae0b.png)

But it wasn't in my feed, even after hard refresh:
![my-feed](https://user-images.githubusercontent.com/10965479/145696854-9a53a933-a8c3-48be-8c92-f41b3e6f02c6.png)

So I restarted Invidious, and there we go - the new video appeared.
![my-feed-restart](https://user-images.githubusercontent.com/10965479/145696819-a50c76cb-d20e-4b5d-b7cf-04c8b04fb071.png)

So, the channel page knew about the new video, but my subscriptions page didn't. I'm running with 16 channel and feed threads, and while we have talked about `use_pubsub_feeds`, this is disabled in this instance for the time being as I still have to do some more work on that.

It's a bit strange that part of the system knows about the video, and another part doesn't - As I mentioned, I've got 101 subs and I benchmarked it in the past and it refreshed in seconds, so I'm not sure what's up here. Any ideas?

**Comment by SamantazFox** [comment]
@hugoamvieira thanks a lot for your investigations! My guess is on one of the two following things: either the non-pubsub "refresh channels" job hangs and is unable to properly update data, or some condition is encountered which prevents the video's metadata to be updated.

In all cases, I think that we need to store the type of the video (is it a live, a premiere, a "shorts", a clip, etc) as it's starting to cause issues now that many different kinds of videos exist across youtube.

**Comment by hugoamvieira** [comment]
@SamantazFox Agreed - Another observation I had, but I think this is more of a feature request is that when livestreams go live / premieres start, they don't get bumped up to the top of the sub feed, so you end up missing the content if it was scheduled days prior.
In terms of job errors, I have my instance set to log anything past `warn` and I see no errors in the logs so presumably it's either something hidden and it's not logged, or it's just a flat out hang that only a restart can fix.

While I'm not fluent at all with Crystal, I did notice that the job that (I think) updates the feeds [doesn't seem to have much in the way of error logging calls](https://github.com/iv-org/invidious/blob/master/src/invidious/jobs/refresh_feeds_job.cr), unless that happens elsewhere / automatically (?)

**Comment by SamantazFox** [comment]
> While I'm not fluent at all with Crystal, I did notice that the job that (I think) updates the feeds [doesn't seem to have much in the way of error logging calls](https://github.com/iv-org/invidious/blob/master/src/invidious/jobs/refresh_feeds_job.cr), unless that happens elsewhere / automatically (?)

You're right, it's really missing verbosity. That's one of the points we have to improve in invidious, in addition to proper tests (to ensure no regressions are implemented in the code when bugs are fixed or features implemented).

Also, now that you point to the source code, I see that the "refresh feeds" job uses materialized views, which are known to cause issues. We have a PR ready to remove that (#2469), we first need a proper migration framework first.

**Comment by SamantazFox** [comment]
Just realized what the problem is: livestreams are treated like any other video, and hence are cached in the DB. We should add some logic to explicitely drop the DB cache entry if the video type is livestream/premiere and is planned to play soon (e.g in the next hour or so) or is currently playing (i.e the stream isn't a normal video yet).

So that rejoins my point above: we really need separation of video types.

**Comment by SamantazFox** [comment]
Switching to a feature request, as there is currently no logic on the client side to regularly check if a live stream has started.
